### ISSUE TITLE
tf.data.Dataset.shuffle produces the same results at each dataset iteration in tensorflow 2 alpha

### ISSUE THREAD

**Issue by fabio12345** [tensorflow/tensorflow#27680]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
Yes. Create a simple dataset, shuffle it and iterate through it.  
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
Ubuntu 18.04
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): 
2.0.0-alpha0
- Python version:
Python 3.7.3
- CUDA/cuDNN version:
CUDA10.0
Not relevant.
- GPU model and memory:
Not relevant.

**Describe the current behavior**
when using tf.data.Dataset.shuffle and iterating through the dataset multiple times the shuffled order is always the same.

**Describe the expected behavior**
The order should change for every iteration.

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.
```
import tensorflow as tf
input_array = tf.range(10)
print(input_array)
dataset = tf.data.Dataset.from_tensor_slices(input_array).shuffle(5)

def print_values():
    for val in dataset:
        print(val.numpy(), end=" ")
    print()

print_values()
print_values()
```

output:
```
tf.Tensor([0 1 2 3 4 5 6 7 8 9], shape=(10,), dtype=int32)
2 1 4 6 0 8 9 5 7 3 
2 1 4 6 0 8 9 5 7 3 

```

**Other info / logs**
Behaviour is correct in tensorflow 1.13.

**Comment by ymodak** [comment]
Thanks for trying TF 2.0 alpha. I was able to reproduce your behavior in TF 2.0 alpha however it executed successfully using TF 1.13.1 (after enabling eager mode) .

**Comment by jsimsa** [comment]
@fabio12345 when you say the behavior is "correct" in TF 1.13, are you talking about graph mode or eager mode? Could you provide an example of the program that you ran with TF 1.13?

When I try the following eager mode program against TF nightly, I see the same behavior as you do with TF 2.0:

```
from __future__ import print_function
import tensorflow as tf

tf.enable_eager_execution()

input_array = tf.range(10)
print(input_array)
dataset = tf.data.Dataset.from_tensor_slices(input_array).shuffle(5)

def print_values():
  for val in iter(dataset):
    print(val.numpy(), end=" ")
  print()

print_values()
print_values()
```

output:
```
tf.Tensor([0 1 2 3 4 5 6 7 8 9], shape=(10,), dtype=int32)
0 5 3 2 4 8 9 1 7 6 
0 5 3 2 4 8 9 1 7 6 
```

The behavior you are seeing is expected (both for TF 2.0 and TF 1.13 eager mode) because your program creates multiple separate iterators and the order of iteration of each is deterministic.

To see different order of elements for each epoch, you can make use the `.repeat()` transformation:
```
from __future__ import print_function
import tensorflow as tf

tf.enable_v2_behavior()

input_array = tf.range(10)
print(input_array)
dataset = tf.data.Dataset.from_tensor_slices(input_array).shuffle(5).repeat(2)

def print_values():
  for val in iter(dataset):
    print(val.numpy(), end=" ")
  print()

print_values()
```

output:
```
tf.Tensor([0 1 2 3 4 5 6 7 8 9], shape=(10,), dtype=int32)
0 2 6 5 7 8 9 1 3 4 2 5 1 3 6 8 9 0 7 4 
0 2 6 5 7 8 9 1 3 4 2 5 1 3 6 8 9 0 7 4 
```

**Comment by jvishnuvardhan** [comment]
@fabio12345 Could you try this

!pip install tensorflow==2.0.0-alpha0
import tensorflow as tf
input_array = tf.range(10)
print(input_array)
dataset = tf.data.Dataset.from_tensor_slices(input_array)
print(dataset)


def print_values():
  for val in dataset:
    print(val.numpy(), end=" ")

for i in range(3):
  dataset=dataset.shuffle(5,reshuffle_each_iteration=True)
  print_values()
  print()

Thanks!

**Comment by edoardogiacomello** [comment]
I'm also experiencing this problem. [API Docs](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/data/Dataset#shuffle) state clearly that the intended behavior is that the dataset is re-shuffled "the dataset should be pseudorandomly reshuffled each time it is iterated over".

In my case the reshuffle_each_iteration parameter has no effect (tried either True, False and None), subsequent iterations of the dataset always lead the same result.

The only workaround that came to my mind is to create a lambda which calls the dataset creation function before every iteration, but I worry this would be inefficient (and not elegant):

` 
data = np.arange(100)
data = data.reshape([10, 10])
def read_dataset():
    return tf.data.Dataset.from_tensor_slices(data).shuffle(100, reshuffle_each_iteration=True).batch(2)
dataset = read_dataset()
dataset_shuf = lambda: read_dataset()
# This doesn't work
for d in dataset:
    tf.print(d)

# This Works
for d in dataset_shuf():
    tf.print(d)

`

@jsimsa If the behavior would be expected as you say, then there's little use in having a "reshuffle_each_iteration" parameter in a framework that is executing eagerly by default. Moreover, repeat() is not a viable option if you need to alternate different type of computations at each dataset, because every time you iterate again it would lead the same sequence of samples.

**Comment by jsimsa** [comment]
@edoardogiacomello 

I agree with you that `reshuffle_each_iteration` in eager mode does not have the expected effect. The documentation is misleading because "iteration" in the documentation is meant to imply repetition based on `repeat` and not based on Python iteration.

Your example of "This Works" works because a new dataset graph is created for each iteration. The value of `reshuffle_each_iteration` is still not respected -- setting `reshuffle_each_iteration` to `False` in your input pipeline will not result in identical order of batches between iterations.

I will create a PR that updates the documentation and renames the argument to `reshuffle_on_repeat` to better communicate what the argument does.

Finally, to achieve the desired effect of reshuffling across different Python iterations of a dataset, your approach based on a dataset factory is idiomatic:

```
def make_dataset():
  return tf.data.Dataset.from_tensor_slices(data).shuffle(100).batch(2)

# illustrate that batches are shuffled in a different order
for _ in range(2):
  for d in make_dataset():
    tf.print(d)
```

**Comment by omoindrot** [comment]
It would be really great to have a way to iterate multiple times over a dataset with different orders.

In the tutorials for TensorFlow 2.0, the recommended way to train a model is:
```python
dataset = ...
dataset = dataset.shuffle(buffer_size)
dataset = dataset.batch(batch_size)

for epoch in range(num_epochs):
    for x, y in dataset:
        train_step(model, x, y, loss_fn)
```

The issue with this code is that at each epoch, the model sees the batches in the same order.

It would be very useful to have a parameter like `reshuffle_each_iteration`, but that would work for each time we iterate over the dataset in eager mode.

---
The current solution described above involves creating a new dataset each time, which doesn't seem ideal:

```python
dataset = ...


for epoch in range(num_epochs):
    dataset_epoch = dataset.shuffle(buffer_size)
    dataset_epoch = dataset_epoch.batch(batch_size)
    for x, y in dataset_epoch:
        train_step(model, x, y, loss_fn)
```

**Comment by jsimsa** [comment]
We are actively working on this and plan to have this ready by TF 2.0 RC0. Most likely, the mechanism for controlling this will be a `tf.data.Option` option (as opposed to `shuffle` option). cc @rohan100jain

**Comment by omoindrot** [comment]
For future reference, this behavior was fixed in this commit: e7206a7d8efceda25bce5a530fc4e79a5582be13

Now when `reshuffle_each_iteration=True`, iterating over the dataset multiple times in eager mode will work as expected.

**Comment by omoindrot** [comment]
@jsimsa : could it be possible to include an example explaining the behavior of `tf.data.Dataset.shuffle` in eager mode with tf 2.0 in the documentation?

The current documentation only mention:
>reshuffle_each_iteration: (Optional.) A boolean, which if true indicates that the dataset should be pseudorandomly reshuffled each time it is iterated over. (Defaults to True.)

which doesn't really explain in full length the new behavior

**Comment by ff4456** [comment]
> We are actively working on this and plan to have this ready by TF 2.0 RC0. Most likely, the mechanism for controlling this will be a `tf.data.Option` option (as opposed to `shuffle` option). cc @rohan100jain

Dear @jsimsa , I am working on a project that requires me to balance the classes in each batch after the dataset is being shuffled. Ideally, I don't want to create a new dataset at the beginning of each epoch. At first I considered the following approach:
```python
def batch_balance(input: tf.data.Dataset):
    #create a function that can re-organize the elements from the shuffled dataset
    #and return a new one that is class-balanced within each batch but still random (to some extent)
    return a_new_dataset
tfds = tf.data.Dataset.from_tensor_slices([[0,0,1],[0,0,1],[0,1,0],[0,1,0],[1,0,0],[1,0,0]])
tfds = tfds.shuffle(buffer_size=6, reshuffle_each_iteration=True)
tfds = tfds.apply(batch_balance)
tfds = tfds.batch(3)
list(tfds.as_numpy_iterator())
list(tfds.as_numpy_iterator())
```
and I get the following output:
>[0,1,0], [0,0,1], [1,0,0]
>[1,0,0], [0,0,1], [0,1,0] (first iteration)

>[0,1,0], [0,0,1], [1,0,0]
>[1,0,0], [0,0,1], [0,1,0] (second iteration)

This does the trick for the first pass, but in fact, after applying the `batch_balance` function, the returned dataset no longer has the` reshuffle_each_iteration` feature, which leads to two identical sequences.

To preserve this reshuffle feature, I've considered writing a modified `shuffle` / `batch`, and then I notice these two functions are both implemented in the `gen_dataset_ops` file, which is machine generated (and very difficult to read...). Therefore, I guess the question is, are there any built-in functions that can allow me to modify the dataset after it is being shuffled w/o losing the `reshuffle_each_iteration` feature? You hinted earlier that `tf.data.Option` can be used, would you mind elaborating a bit more on this?

Thank you!

**Comment by aaudiber** [comment]
@ff4456 `reshuffle_each_iteration` should continue to work even after applying additional transformations. Can you share a colab to reproduce the issue?

**Comment by ff4456** [comment]
Hi @aaudiber , sure! Here is the link: https://colab.research.google.com/drive/1gWxyYHhh5dFTmFbZiEyXwlU4SuUyPY-B?usp=sharing

**Comment by aaudiber** [comment]
The function passed to `apply` is only called once, so you're only ever iterating over the shuffled dataset once. That first iteration is stored into `balanced_list`, which gets reused each time you create an iterator over the `tf.data.Dataset.from_generator`.

To fix this, you can update the generator function to re-iterate over the input to `apply` each time. Here is an updated example of your code that works as expected: https://colab.research.google.com/drive/1Xy8rYbrn2wyJVmvn2uFrSOo2AUHdtxuk

**Comment by ff4456** [comment]
> The function passed to `apply` is only called once, so you're only ever iterating over the shuffled dataset once. That first iteration is stored into `balanced_list`, which gets reused each time you create an iterator over the `tf.data.Dataset.from_generator`.
> 
> To fix this, you can update the generator function to re-iterate over the input to `apply` each time. Here is an updated example of your code that works as expected: https://colab.research.google.com/drive/1Xy8rYbrn2wyJVmvn2uFrSOo2AUHdtxuk

ah I see... Thank you so much!

**Comment by engrmz** [comment]
Hi,

How could we reshuffle dataset in every epoch without using eager mode? 
```
import tensorflow.compat.v1 as tf
tf.disable_eager_execution()
```

In my case, data remains in same sequence after every iteration. I need to reshuffle at position **#Reshuffle is required here** .

```
data = [...]
data = data.shuffle(self.batch_size * 2, reshuffle_each_iteration=True)
data = data.batch(self.batch_size)

iterator = tf.data.Dataset.make_initializable_iterator(data)
batch = iterator.get_next()

epoch = 0
with tf.Session() as sess:
     while epoch < 2:
            batch_count = 0
            sess.run(iterator.initializer)
            # Reshuffle is required here 



            while True:  
                try:
                    **batch_val = sess.run(batch)**
                    print("\nEpoch: {} -- Batch {} :  ".format(epoch, batch_count))

                except tf.errors.OutOfRangeError:
                    epoch += 1
                    break
```
Thanks in advance.

**Comment by aaudiber** [comment]
@engrmz To get different orders you can use `data = data.repeat(num_epochs)`, to repeat the dataset `num_epochs` times, with each repetition doing a reshuffle.

**Comment by engrmz** [comment]
> @engrmz To get different orders you can use `data = data.repeat(num_epochs)`, to repeat the dataset `num_epochs` times, with each repetition doing a reshuffle.

Hi @aaudiber,
Thanks for your prompt response. 
**Point 1:** When I use repeat(5), it repeats the dataset 5 time and shuffles it too. Doing this, I can get shuffled data in every batch. My code (as given above) considers this step as epoch 1 (as I am trying to repeat dataset using epoch loop).

**Point 2:** However, after 5 repetitions of whole dataset, in epoch 2, I still get the same sequence (5 repatriations) as in **Point1:** - same as epoch 1.

Is there any way to shuffle dataset after every epoch iteration without using repeat function?

**Comment by aaudiber** [comment]
The TF1 way of doing things is to handle multiple iterations of the dataset by using `repeat()`. Creating a new iterator in each epoch loop is the TF2 way. If you'd like to discuss further please open a separate issue so we can avoid hijacking this one.
